File: apps/server/src/db/shared/safety.ts
import { parse } from 'pgsql-parser'

export interface StatementRisk {
  command: string
  dangerous: boolean
  reasons: string[]
}

export interface SafetyReport {
  statements: StatementRisk[]
  dangerous: boolean
  reasons: string[]
  parseError: string | null
}

// Map the AST statement key to a human command label.
const COMMANDS: Record<string, string> = {
  SelectStmt: 'SELECT',
  InsertStmt: 'INSERT',
  UpdateStmt: 'UPDATE',
  DeleteStmt: 'DELETE',
  TruncateStmt: 'TRUNCATE',
  DropStmt: 'DROP',
  DropdbStmt: 'DROP DATABASE',
  CreatedbStmt: 'CREATE DATABASE',
  AlterTableStmt: 'ALTER TABLE',
  CreateStmt: 'CREATE TABLE',
  CreateSchemaStmt: 'CREATE SCHEMA',
  VariableSetStmt: 'SET',
}

function analyzeNode(key: string, node: Record<string, unknown>): StatementRisk {
  const command = COMMANDS[key] ?? key.replace(/Stmt$/, '').toUpperCase()
  const reasons: string[] = []

  if (key === 'DeleteStmt' && node.whereClause == null) {
    reasons.push('DELETE without a WHERE clause — every row in the table will be deleted.')
  }
  if (key === 'UpdateStmt' && node.whereClause == null) {
    reasons.push('UPDATE without a WHERE clause — every row in the table will be modified.')
  }
  if (key === 'TruncateStmt') {
    reasons.push('TRUNCATE removes all rows from the table(s).')
  }
  if (key === 'DropdbStmt') {
    reasons.push('DROP DATABASE permanently removes an entire database.')
  }
  if (key === 'DropStmt') {
    const removeType = String(node.removeType ?? '')
    const target = removeType.replace(/^OBJECT_/, '').toLowerCase()
    if (['table', 'schema', 'view', 'matview', 'sequence', 'index'].includes(target)) {
      reasons.push(`DROP ${target.toUpperCase()} permanently removes the object and its data.`)
    }
  }
  if (key === 'AlterTableStmt') {
    const cmds = (node.cmds as { AlterTableCmd?: { subtype?: string } }[]) ?? []
    if (cmds.some((c) => c.AlterTableCmd?.subtype === 'AT_DropColumn')) {
      reasons.push('ALTER TABLE … DROP COLUMN permanently removes a column and its data.')
    }
  }

  return { command, dangerous: reasons.length > 0, reasons }
}

/**
 * Whether the SQL is a single SELECT statement and whether it already has a LIMIT.
 * Used to decide if a default LIMIT/OFFSET can be appended for pagination.
 */
export async function inspectSelect(
  sql: string,
): Promise<{ singleSelect: boolean; hasLimit: boolean }> {
  try {
    const parsed = (await parse(sql)) as { stmts?: { stmt?: Record<string, unknown> }[] }
    const stmts = parsed.stmts ?? []
    if (stmts.length !== 1) return { singleSelect: false, hasLimit: false }
    const stmt = stmts[0].stmt
    const key = stmt ? Object.keys(stmt)[0] : undefined
    if (key !== 'SelectStmt') return { singleSelect: false, hasLimit: false }
    const node = stmt![key] as Record<string, unknown>
    return { singleSelect: true, hasLimit: node.limitCount != null }
  } catch {
    return { singleSelect: false, hasLimit: false }
  }
}

/** Parse SQL with the real PostgreSQL parser and flag dangerous statements. */
export async function analyzeSql(sql: string): Promise<SafetyReport> {
  let parsed: { stmts?: { stmt?: Record<string, unknown> }[] }
  try {
    parsed = (await parse(sql)) as typeof parsed
  } catch (e) {
    // Unparseable — let PostgreSQL report the syntax error at execution time.
    return {
      statements: [],
      dangerous: false,
      reasons: [],
      parseError: e instanceof Error ? e.message : String(e),
    }
  }

  const statements: StatementRisk[] = []
  for (const s of parsed.stmts ?? []) {
    const stmt = s.stmt
    if (!stmt) continue
    const key = Object.keys(stmt)[0]
    if (!key) continue
    statements.push(analyzeNode(key, stmt[key] as Record<string, unknown>))
  }

  const reasons = statements.flatMap((s) => s.reasons)
  return { statements, dangerous: reasons.length > 0, reasons, parseError: null }
}

File: crates/geto-core/src/db/shared/safety.rs
use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::db::shared::split::split_statements;

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct StatementRisk {
    pub command: String,
    pub dangerous: bool,
    pub reasons: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SafetyReport {
    pub statements: Vec<StatementRisk>,
    pub dangerous: bool,
    pub reasons: Vec<String>,
    pub parse_error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SelectInspection {
    pub single_select: bool,
    pub has_limit: bool,
}

fn strip_comments_and_strings(sql: &str) -> String {
    let re_comments = Regex::new(r"(?m)(--[^\n]*|#[^\n]*|/\*[\s\S]*?\*/)").unwrap();
    let stripped = re_comments.replace_all(sql, " ");

    let re_strings = Regex::new(r#"('[^']*'|"[^"]*")"#).unwrap();
    re_strings.replace_all(&stripped, "''").to_string()
}

pub fn analyze_statement(stmt: &str) -> StatementRisk {
    let clean = strip_comments_and_strings(stmt);
    let trimmed = clean.trim();

    let re_cmd = Regex::new(r"(?i)^([a-z]+)").unwrap();
    let command = re_cmd
        .captures(trimmed)
        .map(|c| c[1].to_uppercase())
        .unwrap_or_else(|| "UNKNOWN".to_string());

    let mut reasons = Vec::new();

    let re_where = Regex::new(r"(?i)\bWHERE\b").unwrap();

    match command.as_str() {
        "DELETE" => {
            if !re_where.is_match(trimmed) {
                reasons.push("DELETE without a WHERE clause — every row in the table will be deleted.".to_string());
            }
        }
        "UPDATE" => {
            if !re_where.is_match(trimmed) {
                reasons.push("UPDATE without a WHERE clause — every row in the table will be modified.".to_string());
            }
        }
        "TRUNCATE" => {
            reasons.push("TRUNCATE removes all rows from the table(s).".to_string());
        }
        "DROP" => {
            reasons.push("DROP permanently removes the object and its data.".to_string());
        }
        "ALTER" => {
            let re_drop_col = Regex::new(r"(?i)\bDROP\b").unwrap();
            if re_drop_col.is_match(trimmed) {
                reasons.push("ALTER TABLE … DROP permanently removes a column or constraint.".to_string());
            }
        }
        _ => {}
    }

    StatementRisk {
        command,
        dangerous: !reasons.is_empty(),
        reasons,
    }
}

pub fn analyze_sql(sql: &str) -> SafetyReport {
    let stmts = split_statements(sql);
    let mut statements = Vec::new();
    let mut all_reasons = Vec::new();
    let mut dangerous = false;

    for stmt in stmts {
        let risk = analyze_statement(&stmt);
        if risk.dangerous {
            dangerous = true;
            all_reasons.extend(risk.reasons.clone());
        }
        statements.push(risk);
    }

    SafetyReport {
        statements,
        dangerous,
        reasons: all_reasons,
        parse_error: None,
    }
}

pub fn inspect_select(sql: &str) -> SelectInspection {
    let stmts = split_statements(sql);
    if stmts.len() != 1 {
        return SelectInspection {
            single_select: false,
            has_limit: false,
        };
    }

    let stmt = &stmts[0];
    let clean = strip_comments_and_strings(stmt);
    let trimmed = clean.trim();

    let re_select = Regex::new(r"(?i)^SELECT\b").unwrap();
    let single_select = re_select.is_match(trimmed);

    let re_limit = Regex::new(r"(?i)\bLIMIT\b").unwrap();
    let has_limit = re_limit.is_match(trimmed);

    SelectInspection {
        single_select,
        has_limit,
    }
}
